### PR TITLE
diff:  fix a bug about nil, and can use unique key as order by key

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
1. return null information in ScanRow function
2. can use unique key as order by key

### CODE REVIEW

**Comment by WangXiangUSTC** [comment]
@holys  @GregoryIan  PTAL

File: sync_diff_inspector/diff.go [hunk 1]
```diff
@@ -404,13 +409,16 @@ func (df *Diff) WriteSqls() {
...
-func generateDML(tp string, data map[string][]byte, keys []*model.ColumnInfo, table *model.TableInfo, schema string) (sql string) {
-	// TODO: can't distinguish NULL between ""
+func generateDML(tp string, data map[string][]byte, null map[string]bool, keys []*model.ColumnInfo, table *model.TableInfo, schema string) (sql string) {
 	switch tp {
 	case "replace":
 		colNames := make([]string, 0, len(table.Columns))
 		values := make([]string, 0, len(table.Columns))
 		for _, col := range table.Columns {
+			// don't need add column name and value for NULL
+			if null[col.Name.O] {
```
Comment (holys) [inline sync_diff_inspector/diff.go]: Should we set `NULL` value explicitly, not just omit it?

File: sync_diff_inspector/diff.go [hunk 2]
```diff
@@ -423,6 +431,11 @@ func generateDML(tp string, data map[string][]byte, keys []*model.ColumnInfo, ta
 	case "delete":
 		kvs := make([]string, 0, len(keys))
 		for _, col := range keys {
+			if null[col.Name.O] {
+				kvs = append(kvs, fmt.Sprintf("%s is NULL", col.Name.O))
```
Comment (holys) [inline sync_diff_inspector/diff.go]: Should we use backquote for column name?

**Comment by holys** [comment]
rest LGTM

File: pkg/dbutil/index.go [hunk 3]
```diff
@@ -120,19 +120,30 @@ func SelectUniqueOrderKey(tbInfo *model.TableInfo) ([]string, []*model.ColumnInf
...
 		if index.Primary {
+			keys = keys[:0]
+			keyCols = keyCols[:0]
+			for _, indexCol := range index.Columns {
+				keys = append(keys, indexCol.Name.O)
+				keyCols = append(keyCols, tbInfo.Columns[indexCol.Offset])
+			}
+			break
+		}
+		if index.Unique {
```
Comment (IANTHEREAL) [inline pkg/dbutil/index.go:131]: Do u want to find all unique keys?  I think one unique key (contains multiple columns) is ok

File: pkg/dbutil/query.go [hunk 4]
```diff
@@ -28,10 +28,10 @@ func ScanRowsToInterfaces(rows *sql.Rows) ([][]interface{}, error) {
 }
 
 // ScanRow scans rows into a map.
-func ScanRow(rows *sql.Rows) (map[string][]byte, error) {
+func ScanRow(rows *sql.Rows) (map[string][]byte, map[string]bool, error) {
```
Comment (IANTHEREAL) [inline pkg/dbutil/query.go]: Add some comments for return values?

File: sync_diff_inspector/diff.go [hunk 5]
```diff
@@ -302,19 +302,24 @@ func (df *Diff) checkChunkDataEqual(checkJobs []*CheckJob, table *TableCheckCfg)
...
+		null1, null2 map[string]bool
+		data1, data2 map[string][]byte
+		rowsData1    = make([]map[string][]byte, 0, 100)
+		rowsData2    = make([]map[string][]byte, 0, 100)
+		err          error
+	)
 
 	for rows1.Next() {
-		data1, err := dbutil.ScanRow(rows1)
+		data1, null1, err = dbutil.ScanRow(rows1)
```
Comment (IANTHEREAL) [inline sync_diff_inspector/diff.go]: Don’t need to update null1? Only update rowsData1?

File: sync_diff_inspector/diff.go (hunk 1 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: done

File: sync_diff_inspector/diff.go (hunk 2 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: done

File: pkg/dbutil/index.go (hunk 3 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/index.go:131]: now only use one

File: sync_diff_inspector/diff.go (hunk 5 above)
Comment (WangXiangUSTC) [reply sync_diff_inspector/diff.go]: my mistake, already done

File: pkg/dbutil/query.go (hunk 4 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/query.go]: done

**Comment by WangXiangUSTC** [comment]
@holys @GregoryIan  PTAL again

File: pkg/dbutil/index.go [hunk 6]
```diff
@@ -120,19 +120,32 @@ func SelectUniqueOrderKey(tbInfo *model.TableInfo) ([]string, []*model.ColumnInf
 
 	for _, index := range tbInfo.Indices {
 		if index.Primary {
```
Comment (IANTHEREAL) [inline pkg/dbutil/index.go:122]: `if index.Primary || index.Unique `

**Comment by IANTHEREAL** [comment]
LGTM.

**Comment by holys** [comment]
LGTM

File: pkg/dbutil/index.go (hunk 6 above)
Comment (WangXiangUSTC) [reply pkg/dbutil/index.go:122]: I preferred to use primary key  as order by key when had primary key and unique key

**Pull request merged**